### PR TITLE
Fix Parquet LIST/MAP wrapper extraction in native and Avro readers

### LINKED ISSUE #17420: ParquetNativeRecordReader and ParquetAvroRecordReader expand a json struct MV element as a MV element of maps
Eg: If parquet data is a struct and has below data

```
"metadata": {"tags":["abc", "xyz"]}
```

The parquet readers ParquetNativeRecordReader and ParquetAvroRecordReader expand it to

```
"metadata": {"tags":[{"element":"abc"}, {"element":"xyz"}]}
```

### PR DESCRIPTION
**Pull request by xiangfu0**
## Problem

The native (`ParquetNativeRecordReader`) and Avro-backed (`ParquetAvroRecordReader`) Parquet readers were leaking the LIST/MAP wrapper structs into Pinot rows. A column of `array<string>` came back as `[{"element":"abc"}, {"element":"xyz"}]` instead of `["abc", "xyz"]`, and a `map<string,string>` came back as `{"key_value": [{"key":"k","value":"v"}]}` instead of `{"k":"v"}`. The broken shape propagated into segment generation, multi-value scans, and JSON paths.

Fixes #17420.

### Before
```json
{
  "topLevelTags": [{"element": "top-a"}, {"element": "top-b"}],
  "topLevelProperties": {"key_value": [{"key": "k-a", "value": "v-a"}, {"key": "k-b", "value": "v-b"}]},
  "metadata": {
    "element": "real-element-field",
    "tags": [{"element": "abc"}, {"element": "xyz"}]
  }
}
```

### After
```json
{
  "topLevelTags": ["top-a", "top-b"],
  "topLevelProperties": {"k-a": "v-a", "k-b": "v-b"},
  "metadata": {
    "element": "real-element-field",
    "tags": ["abc", "xyz"]
  }
}
```

## Fix

Drive wrapper detection from the **file schema**, not from the extracted row data.

**`ParquetNativeRecordExtractor`** — use Parquet `LogicalTypeAnnotation` to tell LIST and MAP groups apart from plain structs.
- `extractList`: hoist the per-row `isListElementWrapper` check out of the loop and dispatch the whole list down a single branch. Handles the standard 3-level encoding **and** legacy 2-level encodings (repeated primitive, repeated multi-field group, repeated single-field group not named `element`).
- `extractKeyValueMap`: resolve the `key`/`value` field indices once from the schema and reuse them for every entry.

**`ParquetAvroRecordExtractor`** — check the Avro field schema first, then recurse via `transformValue` on the inner element field so nested `LIST<LIST<...>>` wrappers and `handleDeprecatedTypes` (e.g. INT96 → long) keep applying at every level.

Real struct fields named `element` are preserved; only fields the external schema actually identifies as LIST/MAP wrappers are normalized.

## Note on MAP order

Parquet does **not** guarantee MAP entries are returned in any particular order on read (neither sorted nor insertion order). Writers, page boundaries, and dictionary encodings can all reorder entries. If you need a stable order, use `LIST<STRUCT<key, value>>` instead of the native MAP logical type. This is documented inline on `extractKeyValueMap`.

## Tests

`ParquetCollectionRecordReaderTest` covers:
- Hand-authored Parquet schemas through both readers.
- Avro-schema-written Parquet files through both readers.
- A checked-in golden Parquet fixture (`collection-reader-fixture.parquet`) with primitive types, `DECIMAL`/`DATE`/`TIMESTAMP` logical types, nested structs, LIST and MAP of scalars, struct lists, empty collections, and a real struct field named `element`.
- Legacy 2-level list encodings (single-field non-`element`, multi-field, single-field-named-`element` wrapping a struct).
- Nullable list elements.
- Nested `LIST<LIST<STRING>>` through the Avro reader, including null inner element and null inner wrapper.

## Validation

- `./mvnw -pl pinot-plugins/pinot-input-format/pinot-parquet test` — 19/19 pass.
- `./mvnw spotless:apply checkstyle:check license:check -pl pinot-plugins/pinot-input-format/pinot-parquet` — clean.

### CODE REVIEW

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+
+/**
+ * Shared helpers for normalizing Parquet list wrapper records while extracting Pinot rows.
+ *
+ * <p>This class is stateless and thread-safe.
+ */
+final class ParquetRecordExtractorUtils {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: (minor) Make it public

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  private static final String LIST_ELEMENT_FIELD_NAME = "element";
+  private static final String MAP_ENTRIES_FIELD_NAME = "key_value";
+  private static final String MAP_KEY_FIELD_NAME = "key";
+  private static final String MAP_VALUE_FIELD_NAME = "value";
+
+  private ParquetRecordExtractorUtils() {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: (nit) Put it on the top

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    if (values.length == 0) {
+      return values;
+    }
+
+    boolean hasElementMap = false;
+    for (Object value : values) {
+      if (value == null) {
+        continue;
+      }
+      if (!(value instanceof Map)) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: Can we early terminate?

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      hasElementMap = true;
+    }
+    if (!hasElementMap) {
+      return values;
+    }
+
+    Object[] unwrappedValues = new Object[values.length];
+    for (int i = 0; i < values.length; i++) {
+      Object value = values[i];
+      unwrappedValues[i] = value == null ? null : ((Map<?, ?>) value).get(LIST_ELEMENT_FIELD_NAME);
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: Should we simply replace in place?

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    return unwrappedValues;
+  }
+
+  static Map<String, Object> unwrapMapKeyValues(Map<String, Object> map) {
+    if (map.size() != 1 || !map.containsKey(MAP_ENTRIES_FIELD_NAME)) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: Use `get()` to reduce map access. Same for other places

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      return map;
+    }
+
+    Object entries = map.get(MAP_ENTRIES_FIELD_NAME);
+    Object[] entryArray = toEntryArray(entries);
+    if (entryArray == null || entryArray.length == 0) {
+      return new LinkedHashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: Why not returning `Map.of()`?

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    Map<String, Object> unwrappedMap = new LinkedHashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: Do we need to preserve order? What is the original map type?

File: pinot-plugins/pinot-input-format/pinot-parquet/src/test/java/org/apache/pinot/plugin/inputformat/parquet/ParquetCollectionRecordReaderTest.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+
+/**
+ * Tests Parquet collection wrapper handling in the native and Avro-backed readers.
+ *
+ * <p>This test class is stateful only through its temporary directory and is not thread-safe.
+ */
+public class ParquetCollectionRecordReaderTest {
+  private static final String SCHEMA = "message CollectionWrapperExample {"
+      + "optional group topLevelTags (LIST) {"
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/test/java/org/apache/pinot/plugin/inputformat/parquet/ParquetCollectionRecordReaderTest.java]: Fix the format to be easier to read

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java [hunk 9]
```diff
@@ -63,4 +77,22 @@ Object handleDeprecatedTypes(Object value, Schema.Field field) {
     }
     return value;
   }
+
+  private boolean isListElementWrapper(Schema.Field field) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: We should check this first, then convert value accordingly.
Do we need to handle map in this reader?

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java [hunk 10]
```diff
@@ -195,4 +195,38 @@ public Map<String, Object> extractMap(Group group) {
...
+      GroupType keyValueType = keyValueGroup.getType();
+      Object key = extractValue(keyValueGroup, keyValueType.getFieldIndex("key"));
+      Object value = extractValue(keyValueGroup, keyValueType.getFieldIndex("value"));
+      map.put(key.toString(), value);
+    }
+    return map;
+  }
+
+  @Nullable
+  private Object extractListElement(Group group, Type repeatedField, int index) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java]: Given map is always stored under `"key"` and `"value"`, is that the same for list? Do we need this check for list?

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java (hunk 9 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: Done. `transformValue` now checks `isListElementWrapper(field)` first and only converts/recurses on the inner element field accordingly. For nested `LIST<LIST<...>>` and INT96, recursing via `transformValue` keeps applying `isListElementWrapper` and `handleDeprecatedTypes` at every level — added a regression test for that.

Map handling is delegated to parquet-avro: a Parquet `MAP` is surfaced as an Avro `map` and converted by the base `AvroRecordExtractor.convert()` path, so no extra branch is needed here. Documented inline.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java (hunk 10 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java]: Per the Parquet spec MAP is always under `key`/`value`, so the new `extractKeyValueMap` resolves those indices once from the schema and reuses them.

LIST is different: only the standard 3-level encoding has the inner `element` wrapper. Legacy 2-level encodings exist — repeated primitive (`repeated <type> <name>;`), repeated multi-field group, and repeated single-field group not named `element` (e.g. Hive's `_tuple` / `bag` / `item`) — where the repeated field IS the element. We still need the schema check, but it's hoisted out of the per-row loop now (decided once per LIST value, not per element). Each branch documents the scenario it handles.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: The utility class was removed in the refactor — wrapper detection lives directly in `ParquetAvroRecordExtractor` / `ParquetNativeRecordExtractor` now, so there's no shared helper to expose.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java (hunk 2 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: Done — `LIST_ELEMENT_FIELD_NAME` is at the top of `ParquetAvroRecordExtractor` (line 31).

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: The data-probe loop is gone. Wrapper detection now uses the Parquet/Avro schema directly, so there's nothing to terminate early.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java (hunk 4 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: The build-then-rewrite pattern is gone. `extractKeyValueMap` constructs the result Map directly from the Parquet group entries; there's no intermediate Map to replace.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java (hunk 5 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: No more `Map.containsKey` + `Map.get` patterns — the new code resolves Parquet field indices from the schema once and reads `key` / `value` directly.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java (hunk 6 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: Done — `extractKeyValueMap` returns `Map.of()` for the empty case.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java (hunk 7 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordExtractorUtils.java]: After looking into this, Parquet does not guarantee MAP read order — writers, page boundaries, and dictionary encodings can all reorder entries. Documented this on `extractKeyValueMap` and pointed users who need stable order to `LIST<STRUCT<key, value>>`. Returning a `HashMap`, consistent with the rest of the native extractor.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/test/java/org/apache/pinot/plugin/inputformat/parquet/ParquetCollectionRecordReaderTest.java (hunk 8 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/test/java/org/apache/pinot/plugin/inputformat/parquet/ParquetCollectionRecordReaderTest.java]: Test schemas are now built with `String.join("\n", ...)` instead of hand-concatenated strings — much easier to read.

**Review by xiangfu0 — COMMENTED** [review]
Found one high-signal correctness issue; see inline comment.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java [hunk 11]
```diff
@@ -63,4 +99,22 @@ Object handleDeprecatedTypes(Object value, Schema.Field field) {
...
+    Schema schema = getNonNullSchema(field.schema());
+    if (schema.getType() != Schema.Type.ARRAY) {
+      return false;
+    }
+    Schema elementSchema = getNonNullSchema(schema.getElementType());
+    return elementSchema.getType() == Schema.Type.RECORD && elementSchema.getFields().size() == 1
```
Comment (xiangfu0) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: This shape-only check will also match legitimate Avro schemas like `array<record<element:string>>`. In that case parquet-avro is already giving us the real element record here, so `unwrapListElementWrappers()` will silently flatten `[{"element":"x"}]` into `["x"]`. That is a wrong-results regression for existing Parquet written from Avro schemas with a single-field record named `element`; wrapper detection needs an additional discriminator from the Parquet schema/metadata, or at least a regression test proving this shape is not flattened.

Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: Confirmed real regression — wrote a quick test for `array<record<UserTag, fields=[element: string]>>` and watched it flatten to `["user-1"]`. Thanks for catching it.

Added a discriminator on the inner record's name + namespace. parquet-avro names the synthesized wrapper after the Parquet repeated group (`list` per the 3-level convention, `array` legacy) and either gives it no namespace (top-level field) or a numeric-suffix namespace like `list2`/`list3` to disambiguate when the same record name nests inside another record. User-authored Avro records carry user-chosen names and meaningful namespaces, so requiring `name ∈ {list, array}` AND `namespace == null/empty || namespace.matches(name + "\\d+")` distinguishes the two without needing extra Parquet-schema plumbing.

Pushed [9c566f0](https://github.com/apache/pinot/pull/18325/commits/9c566f0ecc) with the discriminator and a regression test `testAvroRecordReaderPreservesUserRecordWithSingleElementField` that round-trips the user-authored shape and asserts the inner records survive as Maps. All 19 module tests still pass.

Remaining edge case: a user authoring `record<list, no namespace, fields=[element: T]>` (i.e. naming their own record `list` without a namespace) would still false-positive. That seems vanishingly unlikely in practice — open to thoughts if you want a stricter check.

Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: Good call — my first attempt at the discriminator was just an Avro shape heuristic (record name `list`/`array` plus namespace pattern) and that's exactly the fragile thing you flagged. Took another pass with the schema-driven approach.

**Now carrying the Parquet schema end-to-end.** `ParquetAvroRecordReader` reads the Parquet `MessageType` from the file footer and hands it to the extractor along with a `hasAvroSchemaInMetadata` flag (from `ParquetUtils.hasAvroSchemaInFileMetadata`). The extractor maintains a stack of parent Parquet groups while traversing nested records / nested LISTs, so for any Avro field it can look up the corresponding `Type` and check `getLogicalTypeAnnotation()` directly.

**Strip rule** (replaces all the shape/name guessing):

1. Parquet `LogicalTypeAnnotation.ListLogicalTypeAnnotation` says the field IS a LIST — definitive, not inferred from Avro shape.
2. The file does NOT have an Avro schema in metadata — when present, parquet-avro materializes records using the user's Avro schema and `record<element: T>` is user data, not a wrapper.
3. The Avro shape still has the wrapper present (no double-strip when parquet-avro happened to strip on its own).

This naturally distinguishes `array<record<UserTag, [element]>>` (case B — Parquet says LIST, but file has Avro metadata, so parquet-avro already gave us user records → don't strip) from the original bug shape (no Avro metadata, parquet-avro left the wrapper → strip).

[`e5c29d8`](https://github.com/apache/pinot/pull/18325/commits/e5c29d8ab8) — replaces the Avro-shape heuristic, removes the `list`/`array` name + namespace check entirely. All 19 module tests still pass including `testAvroRecordReaderPreservesUserRecordWithSingleElementField`.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java [hunk 12]
```diff
@@ -18,25 +18,174 @@
...
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.Type;
 import org.apache.pinot.plugin.inputformat.avro.AvroRecordExtractor;
 import org.apache.pinot.spi.data.readers.RecordExtractorConfig;
 
 
+/**
+ * Extracts Pinot rows from Avro {@link GenericRecord}s materialized by parquet-avro.
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: Suggest switching to markdown style which is more readable. Same for other javadoc

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java [hunk 13]
```diff
@@ -18,25 +18,174 @@
...
+ *
+ * <p>Parquet's standard 3-level LIST encoding is
+ * {@code <list-rep> group <name> (LIST) { repeated group list { <elem> element; } }}. parquet-avro tries to
+ * strip that wrapper while building the Avro records, but it leaves the wrapper in place when the file has no
+ * Avro schema in its metadata and parquet-avro has to reconstruct the Avro schema from the Parquet schema. When
+ * that happens a Parquet field declared as {@code LIST<T>} surfaces as Avro {@code array<record<element: T>>},
+ * which is the bug fixed here (issue #17420).
+ *
+ * <p>The Avro shape alone is ambiguous — a user-defined Avro {@code array<record<UserTag, [element: string]>>}
+ * matches the same shape and must NOT be flattened. Two pieces of information together resolve the ambiguity:
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: Take a look at `AvroUtils.extractSupportedSchema()`, seems lke we only support single entry record, and directly access the inner value. We should keep the same behavior here

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java [hunk 14]
```diff
@@ -173,26 +176,86 @@ public static long convertInt96ToLong(byte[] int96Bytes) {
   }
 
   public Object[] extractList(Group group) {
-    int numValues = group.getType().getFieldCount();
-    Object[] array = new Object[numValues];
-    for (int i = 0; i < numValues; i++) {
-      array[i] = extractValue(group, i);
-    }
-    if (numValues == 1 && array[0] == null) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java:181]: Is this useful? Can we remove it safely?

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java (hunk 12 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: Done — converted to `///` markdown style on `ParquetAvroRecordExtractor` and `ParquetNativeRecordExtractor` class javadoc. The remaining inline `//` comments inside method bodies are kept as plain comments since they document local control-flow rather than public API.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java (hunk 13 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetAvroRecordExtractor.java]: Good pointer — `AvroUtils.extractSupportedSchema()` recursively unwraps a single-entry RECORD to its inner field's schema (with a `Preconditions.checkState(recordFields.size() == 1)`). That's the schema-inference equivalent of stripping the parquet-avro LIST wrapper at runtime.

In the latest revision ([`b5d171c`](https://github.com/apache/pinot/pull/18325/commits/b5d171cf3b)) the runtime stripping is moved into parquet-avro itself by setting `parquet.avro.add-list-element-records=false` in the Hadoop config (matches Apache Arrow's Parquet behavior). With that flag, parquet-avro flattens the standard 3-level LIST encoding to Avro `array<elem-type>` directly, so the extractor never sees a wrapper record at all — `transformValue` reduces to plain `convert(value)` plus the existing INT96 promotion, no record-handling logic needed. The end-to-end result is consistent with `extractSupportedSchema`'s single-entry assumption (and with the Avro reader's behavior for the same logical input).

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java (hunk 14 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java:181]: Confirmed — the original `extractList` in master had:
```java
public Object[] extractList(Group group) {
  int numValues = group.getType().getFieldCount();   // <- wrong: counts schema fields, not list entries
  Object[] array = new Object[numValues];
  for (int i = 0; i < numValues; i++) {
    array[i] = extractValue(group, i);
  }
  if (numValues == 1 && array[0] == null) {
    return new Object[0];
  }
  return array;
}
```

That implementation walked the LIST group's *schema* fields (which is always 1 for a LIST — the repeated wrapper) instead of the actual repetition count, so it could only ever return a single-element array. The trailing `numValues == 1 && array[0] == null` shortcut covered for the empty-list case where `extractValue` returned null. Both pieces are obsolete with the new spec-driven implementation that uses `group.getFieldRepetitionCount(0)` and dispatches via `isListElementWrapper` per the Parquet LogicalTypes backward-compat rules — empty lists naturally yield `new Object[0]`, no special-casing needed.

Safe to remove ✓ — done in [`b5d171c`](https://github.com/apache/pinot/pull/18325/commits/b5d171cf3b).

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java [hunk 15]
```diff
@@ -173,26 +176,101 @@ public static long convertInt96ToLong(byte[] int96Bytes) {
...
+    for (int i = 0; i < numValues; i++) {
+      Group keyValueGroup = group.getGroup(0, i);
+      Object key = extractValue(keyValueGroup, keyIndex);
+      Object value = extractValue(keyValueGroup, valueIndex);
+      map.put(key.toString(), value);
+    }
+    return map;
+  }
+
+  private boolean isListElementWrapper(Type repeatedField, String parentListName) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java]: Rename this method

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java (hunk 15 above)
Comment (xiangfu0) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetNativeRecordExtractor.java]: Renamed `extractMap` → `extractStruct`. The method handles plain Parquet groups (no LIST/MAP annotation) — i.e. nested struct fields — and returns a Map keyed by child field name. The MAP-annotated path stays as `extractKeyValueMap`, so the two methods now read clearly. Pushed in [`a3d401a`](https://github.com/apache/pinot/pull/18325/commits/a3d401a578).

**Comment by xiangfu0** [comment]
Documentation PR created: https://github.com/pinot-contrib/pinot-docs/pull/765

This PR documents the fix for LIST/MAP wrapper extraction in the Parquet readers, including:
- Explanation of corrected behavior for array and map fields
- Support for 3-level and legacy 2-level Parquet LIST encodings
- **Backward incompatibility warning** for users who worked around the previous broken shape
- MAP order caveat for Parquet readers

Ready for review and merge.
